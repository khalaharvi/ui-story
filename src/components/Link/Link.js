import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styled, { css, cx as emotionCx } from 'react-emotion';

import { createClassNames } from 'utils';

const cx = createClassNames('link');

const isSpecialClick = (event) => {
  const isMiddleClick = event.button === 1;
  return Boolean(
    isMiddleClick ||
      event.altKey ||
      event.ctrlKey ||
      event.metaKey ||
      event.shiftKey
  );
};

export const StyledLink = styled('a')({
  textDecoration: 'none'
});

const Link = ({
  className,
  onClick,
  to,
  children,
  external,
  noFollow
}) => {
  const clickHandler = (e) => {
    if (isSpecialClick(e)) {
      return;
    }
    onClick();
    e.preventDefault();
  };

  let rel = [];
  let target;

  if (external) {
    rel = ['external', 'noopener'];
    target = '_blank';
  }

  if (noFollow) {
    rel.push('nofollow');
  }

  const attrs = {
    rel: rel.length > 0 ? rel.join(' ') : undefined,
    target
  };

  return (
    <StyledLink
      href={to}
      className={classnames(cx(''), className)}
      onClick={onClick ? clickHandler : null}
      {...attrs}
    >
      {children}
    </StyledLink>
  );
};

Link.propTypes = {
  className: PropTypes.string,
  to: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node
  ]).isRequired,
  onClick: PropTypes.func,
  external: PropTypes.bool,
  noFollow: PropTypes.bool
};

Link.defaultProps = {
  external: false,
  noFollow: false
};

const linkWithIcon = css({
  display: 'flex',
  alignItems: 'center',

  'span:first-child': {
    display: 'inline-block',
    width: '17px',
    marginRight: '12px'
  },

  svg: {
    fill: 'currentColor',
    display: 'block',
    margin: '0 auto'
  }
});

export const LinkWithIcon = ({ icon, label, ...props }) => (
  <Link className={emotionCx(cx('-icon'), linkWithIcon)} {...props}>
    <span>{icon}</span>
    <span>{label}</span>
  </Link>
);

LinkWithIcon.propTypes = {
  icon: PropTypes.element.isRequired,
  label: PropTypes.string.isRequired
};

export default Link;
