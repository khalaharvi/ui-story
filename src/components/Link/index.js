export {
  default,
  LinkWithIcon
} from './Link';
