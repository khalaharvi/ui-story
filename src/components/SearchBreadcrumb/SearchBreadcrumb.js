import React from 'react';
import PropTypes from 'prop-types';
import { connectBreadcrumb } from 'react-instantsearch/connectors';
import { cx as emotionCx } from 'react-emotion';
import { withTheme } from 'emotion-theming';

import { createClassNames } from 'utils';
import Link from 'components/Link';

import { getLinkStyle } from 'components/Breadcrumb';
import BreadcrumbList from 'components/Breadcrumb/BreadcrumbList';
import BreadcrumbListItem from 'components/Breadcrumb/BreadcrumbListItem';
import BreadcrumbSeparator from 'components/Breadcrumb/BreadcrumbSeparator';

const cx = createClassNames('breadcrumb');

const itemsPropType = PropTypes.arrayOf(
  PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired
  })
);

// TO DO: add SEO rich markup
export const SearchBreadcrumb = ({
  canRefine,
  createURL,
  items,
  refine,
  rootURL,
  separator,
  alwaysShowRoot,
  rootLabel,
  theme
}) => {
  const link = getLinkStyle(theme);

  const rootPath = (canRefine || alwaysShowRoot) ? (
    <BreadcrumbListItem className={cx('item')}>
      <Link
        className={emotionCx(cx('link'), link)}
        onClick={() => (!rootURL ? refine() : null)}
        to={rootURL || createURL()}
      >
        {rootLabel}
      </Link>
    </BreadcrumbListItem>
  ) : null;

  const breadcrumb = items.map((item, idx) => {
    const isLast = idx === items.length - 1;
    const cleanSeparator = typeof separator === 'string' ? separator.trim() : separator;

    return (
      <BreadcrumbListItem className={cx('item', isLast && 'item--selected')} key={`${item.value}`}>
        <BreadcrumbSeparator className={cx('separator')}>{cleanSeparator}</BreadcrumbSeparator>
        {!isLast ? (
          <Link
            className={emotionCx(cx('link'), link)}
            onClick={() => refine(item.value)}
            to={createURL(item.value)}
          >
            {item.label}
          </Link>
        ) : (
          item.label
        )}
      </BreadcrumbListItem>
    );
  });

  return (
    <div className={cx('', !canRefine && '-noRefinement')}>
      <BreadcrumbList className={cx('list')}>
        {rootPath}
        {breadcrumb}
      </BreadcrumbList>
    </div>
  );
};

SearchBreadcrumb.propTypes = {
  canRefine: PropTypes.bool.isRequired,
  createURL: PropTypes.func.isRequired,
  items: itemsPropType,
  refine: PropTypes.func.isRequired,
  rootURL: PropTypes.string,
  separator: PropTypes.node,
  alwaysShowRoot: PropTypes.bool,
  rootLabel: PropTypes.string
};

SearchBreadcrumb.defaultProps = {
  rootURL: null,
  separator: ' > ',
  alwaysShowRoot: false,
  rootLabel: 'Home'
};

export default connectBreadcrumb(withTheme(SearchBreadcrumb));
