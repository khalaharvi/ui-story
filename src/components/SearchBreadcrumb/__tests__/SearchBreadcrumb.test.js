import React from 'react';
import { mount, shallow } from 'enzyme';

import Link from 'components/Link';
import { SearchBreadcrumb } from '../SearchBreadcrumb';

describe('<SearchBreadcrumb />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <SearchBreadcrumb
        refine={() => null}
        createURL={() => '#'}
        canRefine
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with rootLabel', () => {
    const wrapper = shallow(
      <SearchBreadcrumb
        refine={() => null}
        createURL={() => '#'}
        canRefine
        rootLabel="Homepage"
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with no refinements', () => {
    const wrapper = shallow(
      <SearchBreadcrumb
        refine={() => null}
        createURL={() => '#'}
        canRefine={false}
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with alwaysShowRoot', () => {
    const wrapper = shallow(
      <SearchBreadcrumb
        refine={() => null}
        createURL={() => '#'}
        canRefine={false}
        alwaysShowRoot
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('refines its value on change', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <SearchBreadcrumb
        refine={refine}
        createURL={() => '#'}
        canRefine
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );

    const breadcrumb = wrapper.find('ul');

    expect(breadcrumb.children()).toHaveLength(4);

    breadcrumb
      .children()
      .first()
      .find(Link)
      .simulate('click');
    expect(refine.mock.calls).toHaveLength(1);
    expect(refine.mock.calls[0][0]).toEqual();

    breadcrumb
      .children()
      .at(1)
      .find(Link)
      .simulate('click');
    expect(refine.mock.calls).toHaveLength(2);
    expect(refine.mock.calls[1][0]).toEqual('white');

    breadcrumb
      .children()
      .at(2)
      .find(Link)
      .simulate('click');
    expect(refine.mock.calls).toHaveLength(3);
    expect(refine.mock.calls[2][0]).toEqual('white > white1');

    const lastItem = breadcrumb
      .children()
      .at(3)
      .find(Link);

    expect(lastItem).toHaveLength(0);

    wrapper.unmount();
  });

  it('has a rootURL prop', () => {
    const refine = jest.fn();
    const rootLink = 'deringhall.com';

    const wrapper = mount(
      <SearchBreadcrumb
        refine={refine}
        createURL={() => '#'}
        canRefine
        rootURL={rootLink}
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );

    const breadcrumb = wrapper.find('ul');

    expect(breadcrumb.children()).toHaveLength(4);

    breadcrumb
      .children()
      .first()
      .find(Link)
      .simulate('click');
    expect(refine.mock.calls).toHaveLength(0);
    expect(
      wrapper
        .find('a')
        .first()
        .prop('href')
    ).toEqual('deringhall.com');

    wrapper.unmount();
  });

  it('has a separator prop that can be a custom component', () => {
    const wrapper = shallow(
      <SearchBreadcrumb
        refine={() => null}
        createURL={() => '#'}
        canRefine
        separator={<span role="img" aria-label="separator">üîç</span>}
        items={[
          {
            value: 'white',
            label: 'white'
          },
          {
            value: 'white > white1',
            label: 'white1'
          },
          {
            value: 'white > white1 > white1.1',
            label: 'white1.1'
          }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
