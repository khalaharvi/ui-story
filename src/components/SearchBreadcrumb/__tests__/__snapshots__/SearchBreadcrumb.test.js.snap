// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBreadcrumb /> has a separator prop that can be a custom component 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="dh-breadcrumb"
>
  <Styled(ul)
    className="dh-breadcrumb-list"
  >
    <Styled(li)
      className="dh-breadcrumb-item"
    >
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        Home
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        <span
          aria-label="separator"
          role="img"
        >
          üîç
        </span>
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white > white1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        <span
          aria-label="separator"
          role="img"
        >
          üîç
        </span>
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white1
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item dh-breadcrumb-item--selected"
      key="white > white1 > white1.1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        <span
          aria-label="separator"
          role="img"
        >
          üîç
        </span>
      </Styled(span)>
      white1.1
    </Styled(li)>
  </Styled(ul)>
</div>
`;

exports[`<SearchBreadcrumb /> renders correctly 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="dh-breadcrumb"
>
  <Styled(ul)
    className="dh-breadcrumb-list"
  >
    <Styled(li)
      className="dh-breadcrumb-item"
    >
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        Home
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white > white1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white1
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item dh-breadcrumb-item--selected"
      key="white > white1 > white1.1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      white1.1
    </Styled(li)>
  </Styled(ul)>
</div>
`;

exports[`<SearchBreadcrumb /> renders correctly with alwaysShowRoot 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="dh-breadcrumb dh-breadcrumb--noRefinement"
>
  <Styled(ul)
    className="dh-breadcrumb-list"
  >
    <Styled(li)
      className="dh-breadcrumb-item"
    >
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        Home
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white > white1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white1
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item dh-breadcrumb-item--selected"
      key="white > white1 > white1.1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      white1.1
    </Styled(li)>
  </Styled(ul)>
</div>
`;

exports[`<SearchBreadcrumb /> renders correctly with no refinements 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="dh-breadcrumb dh-breadcrumb--noRefinement"
>
  <Styled(ul)
    className="dh-breadcrumb-list"
  >
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white > white1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white1
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item dh-breadcrumb-item--selected"
      key="white > white1 > white1.1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      white1.1
    </Styled(li)>
  </Styled(ul)>
</div>
`;

exports[`<SearchBreadcrumb /> renders correctly with rootLabel 1`] = `
.emotion-0 {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  className="dh-breadcrumb"
>
  <Styled(ul)
    className="dh-breadcrumb-list"
  >
    <Styled(li)
      className="dh-breadcrumb-item"
    >
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        Homepage
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item"
      key="white > white1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      <Link
        className="dh-breadcrumb-link emotion-0"
        external={false}
        noFollow={false}
        onClick={[Function]}
        to="#"
      >
        white1
      </Link>
    </Styled(li)>
    <Styled(li)
      className="dh-breadcrumb-item dh-breadcrumb-item--selected"
      key="white > white1 > white1.1"
    >
      <Styled(span)
        className="dh-breadcrumb-separator"
      >
        &gt;
      </Styled(span)>
      white1.1
    </Styled(li)>
  </Styled(ul)>
</div>
`;
